skglm/utils.py

```diff
@@ -43,6 +43,69 @@ def box_proj(x, low, up):
         return x
 
 
+@njit
+def value_MCP(w, alpha, gamma):
+    """Compute the value of MCP."""
+    s0 = np.abs(w) < gamma * alpha
+    value = np.full_like(w, gamma * alpha ** 2 / 2.)
+    value[s0] = alpha * np.abs(w[s0]) - w[s0]**2 / (2 * gamma)
+    return np.sum(value)
+
+
+@njit
+def prox_MCP(value, stepsize, alpha, gamma):
+    """Compute the proximal operator of MCP penalty."""
+    tau = alpha * stepsize
+    g = gamma / stepsize  # what does g stand for ?
+    if np.abs(value) <= tau:
+        return 0.
+    if np.abs(value) > g * tau:
+        return value
+    return np.sign(value) * (np.abs(value) - tau) / (1. - 1./g)
+
+
+@njit
+def value_SCAD(w, alpha, gamma):
+    """Compute the value of the SCAD penalty at w."""
+    value = np.full_like(w, alpha ** 2 * (gamma + 1) / 2)
+    for j in range(len(w)):
+        if np.abs(w[j]) <= alpha:
+            value[j] = alpha * np.abs(w[j])
+        elif np.abs(w[j]) <= alpha * gamma:
+            value[j] = (
+                2 * gamma * alpha * np.abs(w[j])
+                - w[j] ** 2 - alpha ** 2) / (2 * (gamma - 1))
+    return np.sum(value)
+
+
+@njit
+def prox_SCAD(value, stepsize, alpha, gamma):
+    """Compute the proximal operator of SCAD penalty."""
+    tau = gamma * alpha
+    x_1 = max(0, np.abs(value) - alpha * stepsize)
+    x_2 = ((gamma - 1) * np.abs(value) - stepsize * tau) / (
+        gamma - 1 - stepsize)
+    x_2 = abs(x_2)
+
+    x_3 = abs(value)
+
+    objs = np.zeros(3)
+    objs[0] = (0.5 / stepsize) * (
+        x_1 - np.abs(value)) ** 2 + value_SCAD(np.array([x_1]), alpha, gamma)
+    objs[1] = (0.5 / stepsize) * (
+        x_2 - np.abs(value)) ** 2 + value_SCAD(np.array([x_2]), alpha, gamma)
+    objs[2] = (0.5 / stepsize) * (
+        x_3 - np.abs(value)) ** 2 + value_SCAD(np.array([x_3]), alpha, gamma)
+
+    idx_min = np.argmin(objs)
+    if idx_min == 0:
```

Comment:
can this be made simpler by first putting x1, x2 and x3 into a list ? then x_list[np.argmin(values])]